runbooks/schedule_backfill.ipynb

```diff
@@ -0,0 +1,612 @@
+    "import json\n",
+    "import re\n",
+    "from collections import Counter\n",
+    "from concurrent.futures import Future, ThreadPoolExecutor, as_completed\n",
+    "from pprint import pprint\n",
+    "from zipfile import ZipFile, ZipInfo\n",
+    "from typing import Dict\n",
+    "from pydantic import BaseModel, ValidationError, parse_obj_as\n",
+    "\n",
+    "# jinja pattern for removing the auth query parameter if it exists\n",
+    "jinja_pattern = r\"(?<=\\?)(?:api_key|token)=[\\w-]+&?\"\n",
+    "\n",
+    "\n",
+    "class SkipUrl(Exception):\n",
+    "    pass\n",
+    "\n",
+    "\n",
+    "class ZipTask(BaseModel):\n",
+    "    outcome: GTFSDownloadOutcome\n",
+    "    gcs_dir: Optional[str]\n",
+    "    files: Optional[Dict[str, pendulum.DateTime]]\n",
+    "\n",
+    "\n",
+    "def feed_folder_to_zip_task(\n",
+    "    folder, feed, zip_one_feed_fs=None, pbar=None\n",
+    ") -> ZipTask:\n",
+    "    zip_one_feed_fs = get_fs() if not zip_one_feed_fs else zip_one_feed_fs\n",
+    "    feed_key = f\"{feed['itp_id']}_{feed['url_number']}\"\n",
+    "    url = re.sub(jinja_pattern, \"\", feed[\"gtfs_schedule_url\"]).rstrip(\"?\")\n",
+    "\n",
+    "    assert (\n",
+    "        url\n",
+    "        and \"token\" not in url\n",
+    "        and \"api_key\" not in url\n",
+    "        and not url.endswith(\"?\")\n",
+    "        and not url.endswith(\"&\")\n",
+    "        and \"?&\" not in url\n",
+    "    )\n",
+    "\n",
+    "    try:\n",
+    "        validated_url = parse_obj_as(HttpUrl, url)\n",
+    "    except ValidationError:\n",
+    "        if url.startswith(\"http://.232\"):\n",
+    "            raise SkipUrl\n",
+    "        raise\n",
+    "\n",
+    "    config = GTFSDownloadConfig(\n",
+    "        extracted_at=None,\n",
+    "        name=feed[\"agency_name\"],\n",
+    "        url=validated_url,\n",
+    "        feed_type=GTFSFeedType.schedule,\n",
+    "        schedule_url_for_validation=None,\n",
+    "        auth_query_params={},\n",
+    "        auth_headers={},\n",
+    "        computed=True,\n",
+    "    )\n",
+    "\n",
+    "    if feed[\"status\"] != \"success\":\n",
+    "        return ZipTask(\n",
+    "            outcome=GTFSDownloadOutcome(\n",
+    "                success=False,\n",
+    "                exception=Exception(feed[\"status\"]),\n",
+    "                config=config,\n",
+    "                extract=None,\n",
+    "                backfilled=True,\n",
+    "            ),\n",
+    "        )\n",
+    "\n",
+    "    feed_folder = f\"{folder}/{feed_key}\"\n",
+    "    files_to_timestamps = {}\n",
+    "\n",
+    "    for current_dir, sub_dirs, files in zip_one_feed_fs.walk(feed_folder):\n",
+    "        # skip our processed subdir\n",
+    "        # skip situations where we have a weird duplicate subdir\n",
+    "        if (\n",
+    "            current_dir.endswith(\"processed\")\n",
+    "            or \"__MACOSX\" in current_dir\n",
```

Comment:
Oh honestly I assumed this was someone fixing data back in v1.

We keep all other subdirs except for these.

Reply:
Ok, in the v2 pipeline we keep MACOSX stuff. See: 
```
SELECT * 
FROM `cal-itp-data-infra.mart_gtfs_quality.fct_schedule_feed_files`
WHERE original_filepath LIKE "%MACOSX%"
```
I don't think it is necessarily caused by us.

Reply:
Looks like it occurs in at least 3 different feeds

Reply:
I doubt it is worth re-running for this, but just noting

Reply:
Ok, weirdly these guys don't have unzip failures in the v2 pipeline? I guess if there is a hidden directory we aren't calling that illegal. Hmmm

Reply:
Just the files are hidden, not the directory. I think we don't encounter it in v2 because we never end up processing those hidden files.

```
gs://test-calitp-gtfs-schedule-unzipped/._agency.txt/
gs://test-calitp-gtfs-schedule-unzipped/._calendar.txt/
gs://test-calitp-gtfs-schedule-unzipped/._calendar_attributes.txt/
gs://test-calitp-gtfs-schedule-unzipped/._calendar_dates.txt/
gs://test-calitp-gtfs-schedule-unzipped/._directions.txt/
gs://test-calitp-gtfs-schedule-unzipped/._fare_attributes.txt/
gs://test-calitp-gtfs-schedule-unzipped/._fare_rules.txt/
gs://test-calitp-gtfs-schedule-unzipped/._feed_info.txt/
gs://test-calitp-gtfs-schedule-unzipped/._realtime_routes.txt/
gs://test-calitp-gtfs-schedule-unzipped/._route_attributes.txt/
gs://test-calitp-gtfs-schedule-unzipped/._routes.txt/
gs://test-calitp-gtfs-schedule-unzipped/._shapes.txt/
gs://test-calitp-gtfs-schedule-unzipped/._stop_times.txt/
gs://test-calitp-gtfs-schedule-unzipped/._stops.txt/
gs://test-calitp-gtfs-schedule-unzipped/._trips.txt/
```

Reply:
The determination about valid structure is made in the unzip job

Reply:
Theoretically if there is a folder and it's not a top level folder that contains the whole feed we should say it's invalid

Reply:
Ah ok I see what you were saying